librubyfmt/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+pub extern "C" fn rubyfmt_init() -> libc::c_int {
+    init_logger();
+    unsafe {
+        ruby::ruby_init();
+    }
+    let res = load_ripper();
+    if res.is_err() {
+        return InitStatus::ERROR as libc::c_int;
+    }
+
+    let res = load_rubyfmt();
+    if res.is_err() {
+        return InitStatus::ERROR as libc::c_int;
+    }
+
+    InitStatus::OK as libc::c_int
+}
+
+pub fn format_buffer(buf: String) -> String {
+    eprintln!("format 1");
+    let bytes: Vec<libc::c_char> = buf
+        .into_bytes()
+        .into_iter()
+        .map(|v| v as libc::c_char)
+        .collect();
+    let len = bytes.len();
+    eprintln!("format 2");
+    let fb = rubyfmt_format_buffer(FormatBuffer {
+        bytes: bytes.as_ptr(),
+        count: len as i64,
+    });
+    eprintln!("format 3");
+    fb.into_string()
+}
+
+#[no_mangle]
+pub extern "C" fn rubyfmt_format_buffer(buf: FormatBuffer) -> FormatBuffer {
+    let output_data = Vec::with_capacity(buf.count as usize);
+    let mut output = Cursor::new(output_data);
+    let tree = run_parser_on(buf);
+    if tree.is_err() {
+        unsafe {
+            let cstr = CString::new("oh no").expect("we just made it");
+            ruby::rb_raise(ruby::rb_eRuntimeError, cstr.as_ptr());
+        }
+    }
+    let tree = tree.expect("we raised");
+    let data = buf.into_buf();
+    let res = toplevel_format_program(&mut output, data, tree);
+    raise_if_error(res);
+    let output_data = output.into_inner().into_boxed_slice();
+    let ptr = output_data.as_ptr();
+    let len = output_data.len();
+    let fb = FormatBuffer {
+        bytes: ptr as *const libc::c_char,
+        count: len as i64,
+    };
+    std::mem::forget(output_data);
+    fb
+}
+
+fn load_rubyfmt() -> Result<VALUE, ()> {
+    let rubyfmt_program = include_str!("../rubyfmt_lib.rb");
+    eval_str(rubyfmt_program)
+}
+
+fn load_ripper() -> Result<(), ()> {
+    // trick ruby in to thinking ripper is already loaded
+    eval_str(
+        r#"
+    $LOADED_FEATURES << "ripper.bundle"
+    $LOADED_FEATURES << "ripper.so"
+    $LOADED_FEATURES << "ripper.rb"
+    $LOADED_FEATURES << "ripper/core.rb"
+    $LOADED_FEATURES << "ripper/sexp.rb"
+    $LOADED_FEATURES << "ripper/filter.rb"
+    $LOADED_FEATURES << "ripper/lexer.rb"
+    "#,
+    )?;
+
+    // init the ripper C module
+    unsafe { Init_ripper() };
+
+    //load each ripper program
+    eval_str(include_str!(
+        "../ruby_checkout/ruby-2.6.6/ext/ripper/lib/ripper.rb"
+    ))?;
+    eval_str(include_str!(
+        "../ruby_checkout/ruby-2.6.6/ext/ripper/lib/ripper/core.rb"
+    ))?;
+    eval_str(include_str!(
+        "../ruby_checkout/ruby-2.6.6/ext/ripper/lib/ripper/lexer.rb"
+    ))?;
+    eval_str(include_str!(
+        "../ruby_checkout/ruby-2.6.6/ext/ripper/lib/ripper/filter.rb"
+    ))?;
+    eval_str(include_str!(
+        "../ruby_checkout/ruby-2.6.6/ext/ripper/lib/ripper/sexp.rb"
+    ))?;
+
+    Ok(())
+}
+
+fn eval_str(s: &str) -> Result<VALUE, ()> {
```

Comment:
One thing I noticed is that you never call this with anything other than `&'static str`. If the input string were always nul terminated you could skip some copies (but that would require either taking `CStr`, an annoying type to construct, or assuming `&str` is always nul terminated, which would make it unsafe). However, since you are only ever taking literals, you could make this a macro and safely do it.

```rust
macro_rules! eval_str {
    ($s:expr) => {
        let mut state = 0;
        ruby::rb_eval_string_protect(
            concat!(s, "\0").as_ptr(),
            &mut state,
        );
        if state != 0 {
            return Err(())
        }
    }
}
```

(note: I omitted the return value since it's never used)